### PR TITLE
Mark members of Event-related types.

### PR DESCRIPTION
**Pull request by erozenfeld**
1. Mark static fields in Keywords, Tasks, and Opcodes classes nested inside EventSource-derived classes.
Details on the pattern relying on this are described here: https://msdn.microsoft.com/en-us/library/dn774985(v=pandp.20).aspx

2. Mark public instance properties inside types with EventDataAttribute.
Details are here: https://msdn.microsoft.com/en-us/library/dn823294(v=vs.110).aspx

### CODE REVIEW

**Comment by erozenfeld** [comment]
@marek-safar PTAL
/cc: @ericstj @sbomer

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -614,6 +621,11 @@ void MarkMethodSpecialCustomAttributes (MethodDefinition method)
 			}
 		}
 
+		void MarkTypeWithEventDataAttribute(TypeDefinition type, CustomAttribute attribute)
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Please follow existing code style

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 2]
```diff
@@ -810,6 +822,29 @@ static bool IsMulticastDelegate (TypeDefinition td)
...
+		{
+			TypeReference type = td;
+			do {
+				if (type.FullName == "System.Diagnostics.Tracing.EventSource") {
+					return true;
+				}
+				TypeDefinition typeDef = type.Resolve ();
+				type = typeDef.BaseType;
+			}
+			while (type != null);
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Needs to be on same line with }

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 3]
```diff
@@ -810,6 +822,29 @@ static bool IsMulticastDelegate (TypeDefinition td)
...
 
+		static bool IsEventSource(TypeDefinition td)
+		{
+			TypeReference type = td;
+			do {
+				if (type.FullName == "System.Diagnostics.Tracing.EventSource") {
+					return true;
+				}
+				TypeDefinition typeDef = type.Resolve ();
+				type = typeDef.BaseType;
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: typeDef can be null here

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 4]
```diff
@@ -810,6 +822,29 @@ static bool IsMulticastDelegate (TypeDefinition td)
...
+			}
+			while (type != null);
+			return false;
+		}
+
+		void MarkEventSource(TypeDefinition td)
+		{
+			foreach (var nestedType in td.NestedTypes) {
+				if ((nestedType.Name == "Keywords") || (nestedType.Name == "Tasks") || (nestedType.Name == "Opcodes")) {
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Extra parens and perhaps switch would look better

File: linker/Mono.Linker.Steps/MarkStep.cs [hunk 5]
```diff
@@ -1136,6 +1182,11 @@ static internal bool IsPropertyMethod (MethodDefinition md)
 				(md.SemanticsAttributes & MethodSemanticsAttributes.Setter) != 0;
 		}
 
+		static internal bool IsPublicInstancePropertyMethod(MethodDefinition md)
```
Comment (marek-safar) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Do you really need it as internal property?

File: linker/Mono.Linker.Steps/MarkStep.cs (hunk 1 above)
Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Done.

File: linker/Mono.Linker.Steps/MarkStep.cs (hunk 3 above)
Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Fixed.

File: linker/Mono.Linker.Steps/MarkStep.cs (hunk 2 above)
Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: Done.

File: linker/Mono.Linker.Steps/MarkStep.cs (hunk 4 above)
Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: I removed the parens but I think a switch would be an overkill here.

File: linker/Mono.Linker.Steps/MarkStep.cs (hunk 5 above)
Comment (erozenfeld) [inline linker/Mono.Linker.Steps/MarkStep.cs]: I don't understand the question. This is an internal method, not a property, just like IsPropertyMethod and IsEventMethod. What's your preference?

**Comment by erozenfeld** [comment]
@marek-safar I addressed the feedback, PTAL.

**Comment by marek-safar** [comment]
Please rebase the change

**Pull request merged**